### PR TITLE
Add rxjs-suffix-subjects rule + tests

### PR DESCRIPTION
**Pull request by maggie-x**
Currently there are options for

- setting the suffix
- whether this rule is enforced for functions, methods, parameters, properties, variables
- whether certain types are exempt from this rule (EventEmitters are exempt by default)

Let me know if you want to options to be organised/categorised in another way.

### CODE REVIEW

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+const defaultTypesRegExp = /^EventEmitter$/;
+
+export class Rule extends Lint.Rules.TypedRule {
+  public static metadata: Lint.IRuleMetadata = {
+    description:
+      "Ensures subjects are suffixed with 'Subject'.",
+    options: {
+      properties: {
+        functions: { type: "boolean" },
+        methods: { type: "boolean" },
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: <s>`functions` and `methods` don't appear to be used - at least not in the test that I read before I read the code. If they're not used, they should be removed from the options.</s>

Never mind. A closer look at the code shows that they are used.

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    sourceFile: ts.SourceFile,
+    program: ts.Program
+  ): Lint.RuleFailure[] {
+    const failures: Lint.RuleFailure[] = [];
+    const typeChecker = program.getTypeChecker();
+
+    // get configurations
+    const { ruleArguments } = this.getOptions();
+    const [options] = ruleArguments;
+    if (options) {
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: This looks to me like the `options` must always be specified in the `tslint.json` configuration. Ideally, the options should be optional - that's what's in the rule's metadata at the top of the file.

Have a look at the implementation of some of the other rules. They should how default options can be applied.

Also, I would write two separate test fixtures: one with the options specified (as in your test); and one in which there are no options specified and the defaults are relied upon.

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    if (options.methods) {
+      identifiers = identifiers.concat(
+        tsquery(
+          sourceFile,
+          `:matches(MethodDeclaration, MethodSignature) > Identifier`
+        )
+      );
+    }
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: I can see that functions and methods are used. So you can disregard the first comment in this review. However, I think functions and methods should be covered in the tests.

Also, I think that maybe the `functions` and `methods` options should default to `false`. What do you think? Do you use a suffix for functions and methods? Given your other rule, I suspect you don't have any functions or methods that return subjects.

**Comment by cartant** [comment]
Another observation: if the suffix is `"Subject"` and there is an identifier named 'subject', will the rule complain and insist upon it being named `subjectSubject`?

I don't think it should.

I've not checked the implementation, so perhaps you've accounted for this? If you have, this needs to be a case that's included in the tests.

**Comment by cartant** [comment]
TBH, I would seriously consider removing the `function` and `method` options. Do you have a use case for them? If you don't, I would suggest you remove them to simplify the rule.

If it just enforces the suffix for variables, parameters and properties (including accessors), I think it's a little easier to grok and explain.

Defaulting the others to `true` is fine.

**Comment by maggie-x** [comment]
Yep I've already removed the `function` and `method` options :+1:

**Comment by maggie-x** [comment]
With the `subjectSubject` issue, I just realised that this brings up more things to consider. 
What if the suffix is `Sbjt` and a subject is called `sbjt`, would this is allowed or would this not be valid? 

I guess the more general rule is if the suffix has a beginning capital letter, and the variable itself is the suffix, but with all letters lowercase, then there should not be a failure?

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  private suffix = 'Subject';
+  private suffixRegex = new RegExp(`${this.suffix}\\$?$`, "i");
+  private types: { regExp: RegExp; validate: boolean }[] = [];
+  private validate = {
+    parameters: true,
+    properties: true,
+    variables: true
+  };
+  public readonly FAILURE_MESSAGE = (identifier: string) =>
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: Given that the rule is no longer using a walker, these private properties don't really need to be class properties. They could be moved into the `applyWithProgram` function. That would be simpler.

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  ): Lint.RuleFailure[] {
+    const failures: Lint.RuleFailure[] = [];
+    const typeChecker = program.getTypeChecker();
+
+    // get configurations
+    const { ruleArguments } = this.getOptions();
+    const [options] = ruleArguments;
+    if (options) {
+      this.suffix = options.suffix;
+      this.suffixRegex = new RegExp(`${this.suffix}\\$?$`, "i");
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: If the private properties are moved into this function - as variables - the regular expression can be created just once: after the `if` statement that checks and applies the options.

File: source/rules/rxjsSuffixSubjectsRule.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      identifiers = identifiers.concat(
+        tsquery(sourceFile, `Parameter > Identifier`)
+      );
+    }
+
+    if (validateOptions.properties) {
+      identifiers = identifiers.concat(
+        tsquery(
+          sourceFile,
+          `:matches(PropertyAssignment, PropertyDeclaration, PropertySignature, GetAccessor, SetAccessor) > Identifier`
```
Comment (cartant) [inline source/rules/rxjsSuffixSubjectsRule.ts]: Very nice selector!

**Comment by cartant** [comment]
Thanks. [4.17.0](https://github.com/cartant/rxjs-tslint-rules/blob/master/CHANGELOG.md#4170-2019-02-08) has been published and it includes your rule.

**Pull request merged**